incrementalmerkletree/src/lib.rs

```diff
@@ -464,6 +464,53 @@ impl<'a> From<&'a Address> for Option<Position> {
     }
 }
 
+/// A path from a position in a particular commitment tree to the root of that tree.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MerklePath<H, const DEPTH: u8> {
+    path_elems: Vec<H>,
+    position: Position,
+}
+
+impl<H, const DEPTH: u8> MerklePath<H, DEPTH> {
+    /// Constructs a Merkle path directly from a path and position.
+    #[allow(clippy::result_unit_err)]
+    pub fn from_parts(path_elems: Vec<H>, position: Position) -> Result<Self, ()> {
+        if path_elems.len() == usize::from(DEPTH) {
+            Ok(MerklePath {
+                path_elems,
+                position,
+            })
+        } else {
+            Err(())
+        }
+    }
+
+    pub fn path_elems(&self) -> &[H] {
+        &self.path_elems
+    }
+
+    pub fn position(&self) -> Position {
+        self.position
+    }
+}
+
+impl<H: Hashable, const DEPTH: u8> MerklePath<H, DEPTH> {
+    /// Returns the root of the tree corresponding to this path applied to `leaf`.
+    pub fn root(&self, leaf: H) -> H {
+        self.path_elems
+            .iter()
+            .enumerate()
+            .fold(leaf, |root, (i, h)| {
+                let level = Level(i as u8);
+                if self.position.0 >> i & 0x1 == 0 {
```

Comment:
```suggestion
                if (self.position.0 >> i) & 0x1 == 0 {
```
It's correct as-is, but this is clearer I think. Non-blocking.